### PR TITLE
Added BIP 69: Lexicographical Indexing of Inputs and Outputs

### PR DESCRIPTION
**Pull request by kristovatlas**
Lexicographical Indexing of Transaction Inputs and Outputs

### CODE REVIEW

**Comment by kristovatlas** [comment]
Electrum 2.3.2 has implemented BIP 79: https://github.com/spesmilo/electrum/blob/master/RELEASE-NOTES

**Comment by laanwj** [comment]
Maybe add reference to discussion on mailing list: https://lists.linuxfoundation.org/pipermail/bitcoin-dev/2015-June/008484.html

**Comment by laanwj** [comment]
Weird: seems to need rebase

**Comment by dcousens** [comment]
LGTM,  anything holding this back?

File: bip-0069.mediawiki [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+
+Because the hash of previous transactions and output indices must be included in a signed transaction, wallet clients capable of signing transactions will necessarily have access to this data.
+
+Transaction malleability will not negatively impact the correctness of this process. Even if a wallet client follows this process using unconfirmed UTXOs as inputs and an attacker changes modifies the blockchain’s record of the hash of the previous transaction, the wallet client will include the invalidated previous transaction hash in its input data, and will still correctly sort with respect to that invalidated hash.
```
Comment (dcousens) [inline bip-0069.mediawiki]: Is this relevant?

File: bip-0069.mediawiki [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+    elif (len(barr1) > len(barr2)):
+        return 1
+    else:
+        return 0</nowiki>
+
+N.B. These comparisons do not need to operate in constant time since they are not processing secret information.
+
+===Transaction Inputs===
+
+104	Transaction inputs are defined by the hash of a previous transaction, the output index of of a UTXO from that previous transaction, the size of an unlocking script, the unlocking script, and a sequence number. [3] For sorting inputs, the hash of the previous transaction and the output index within that transaction are sufficient for sorting purposes; each transaction hash has an extremely high probability of being unique in the blockchain — this is enforced for coinbase transactions by BIP30 — and output indices within a transaction are unique. For the sake of efficiency, transaction hashes should be compared first before output indices, since output indices from different transactions are often equivalent, while all bytes of the transaction hash are effectively random variables.
```
Comment (dcousens) [inline bip-0069.mediawiki]: This entire paragraph is quite verbose and seems more like a primer on things we already know.

File: bip-0069.mediawiki [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+
+Hashes of previous transactions are considered for the purposes of this BIP in their little-endian, byte array format in order to match the traditional, human-readable string representation of the hashes. They must be sorted in accordance with the output of the bytearr_cmp() function above: the hash with the earliest lesser byte is ordered first, and shorter hashes are ordered before longer ones as a tie-breaker. In the event of two matching transaction hashes, output indices will be compared based on their integer value, with the smaller value ordered first. A further tie is extremely improbable for the aforementioned reasons.
```
Comment (dcousens) [inline bip-0069.mediawiki]: > A further tie is extremely improbable for the aforementioned reasons.

This isn't just statistically improbably, but its blatantly not allowed by the network right?

File: bip-0069.mediawiki [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+
+===Applicability===
+
+This BIP applies to any transaction for which the order of its inputs and outputs does not impact the transaction’s function. Currently, this refers to any transaction that employs the SIGHASH_ALL signature hash type, in which signatures commit to the exact order of inputs and outputs. Transactions that use SIGHASH_ANYONECANPAY and/or SIGHASH_NONE may include inputs and/or outputs that are not signed; however, compliant software should still emit transactions with lexicographically sorted inputs and outputs, even though they may later be modified by others.
+
+In the event that future protocol upgrades introduce new signature hash types, compliant software should apply the lexicographical ordering principle analogously.
+
+While out of scope of this BIP, protocols that do require a specified order of inputs/outputs (e.g. due to use of SIGHASH_SINGLE) should consider the goals of this BIP and how best to adapt them to the specific needs of those protocols.
+
+===Lexicographical Sorting===
```
Comment (dcousens) [inline bip-0069.mediawiki]: I'm not sure if an algorithmic description of lexicographical sorting is required, this entire section could be summarised (including the pseudocode) with just a link to the [Lexicographical ordering Wikipedia entry](https://en.wikipedia.org/wiki/Lexicographical_order).

File: bip-0069.mediawiki (hunk 3 above)
Comment (dcousens) [inline bip-0069.mediawiki]: > shorter hashes are ordered before longer ones as a tie-breaker

Transaction hashes are *always* going to be 32 bytes,  as they are padded.   Why are we accounting for 'shorter hashes'?

File: bip-0069.mediawiki [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+
+Because the hash of previous transactions and output indices must be included in a signed transaction, wallet clients capable of signing transactions will necessarily have access to this data.
+
+Transaction malleability will not negatively impact the correctness of this process. Even if a wallet client follows this process using unconfirmed UTXOs as inputs and an attacker changes modifies the blockchain’s record of the hash of the previous transaction, the wallet client will include the invalidated previous transaction hash in its input data, and will still correctly sort with respect to that invalidated hash.
+
+===Transaction Outputs===
+
+A transaction output is defined by its scriptPubKey and amount. [3] For sorting purposes, we will consider a scriptPubKey in its byte array representation, and a bitcoin amount in terms of their integer number of satoshis (smallest amount ordered first).
```
Comment (dcousens) [inline bip-0069.mediawiki]: > smallest amount ordered first

Commonly known as `ascending order`, a nit pick,  but,  it is clearer IMHO.

**Comment by dcousens** [comment]
Test fixtures can be found [here](https://github.com/bitcoinjs/bip69/blob/master/test/fixtures.json).

A Javascript implementation can be found [here](https://github.com/bitcoinjs/bip69/blob/master/index.js), and on [npm](https://www.npmjs.com/package/bip69).

**Comment by dcousens** [comment]
@kristovatlas thoughts on integrating the anti-fee-sniping behaviour (see https://github.com/bitcoin/bitcoin/pull/6216) into BIP69?

**Comment by kristovatlas** [comment]
What's fee sniping?

**Comment by dcousens** [comment]
@kristovatlas to quote @petertodd from https://github.com/bitcoin/bitcoin/pull/2340

> The more important reason is to discourage "fee sniping" by deliberately mining blocks that orphan the current best block. Basically for a large miner the value of the transactions in the best block and the mempool can exceed the cost of deliberately attempting to mine two blocks to orphan the best block. However with nLockTime you'll soon run out of transactions you can put in the first block, which means they now need to go in the second. With limited block sizes you're run out of room, and additionally another miner now only needs to orphan one block to in-turn snipe the high-fee transactions you had to place in the second block, wrecking all your hard work.

How that is relevant to this BIP is that it does mean transactions are more easily recognized due to `nLockTime` being set to `activeChain.Height() + 1`.

**Comment by kristovatlas** [comment]
I can sort of see the relationship there, but I'd prefer to keep this BIP straightforward and restricted to advice to Bitcoin wallet clients.

**Comment by kristovatlas** [comment]
> Test fixtures can be found here.
>
> A Javascript implementation can be found here, and on npm.

Thanks a lot! I will add links to these in the BIP.

**Comment by kristovatlas** [comment]
I tried rebasing my local copy but to no avail, so I'm going to try submitting a new PR...

**Comment by kristovatlas** [comment]
Looks like I managed to resolve them. Please consider merging, @laanwj

**Comment by dcousens** [comment]
@kristovatlas any chance of addressing the comments I made above? (on the diff)

**Comment by kristovatlas** [comment]
@dcousens: Thanks for your feedback. I already spent a long time editing the BIP based on prior feedback and I'm not really interested in spending more editing for the sake of succinctness, which comes with trade-offs. When I interacted with open source contributors to wallets on this topic, I found that they primarily wanted more detail about the scheme and not less. Keep in mind that the audience for BIPs is not necessarily restricted to people who religiously read the dev mailing list -- not that you are suggesting this, it's just a helpful reference point.

**Comment by dcousens** [comment]
@kristovatlas if possible,  I'd prefer to stay away from heresay,  as,  in my experience,  I've only heard the opposite in regards to developer opinion on this BIP. 

The fact is https://github.com/bitcoin/bips/pull/157#discussion_r36488055 does not make any sense [to me],  and although it seems well intentioned,  I can't possibly imagine why you have accounted for it?

The BIP is currently a very long explanation for the following algorithm (pseudocode):
``` javascript
inputs.sort((a, b) => compare(a.txHash, b.txHash, 32) || a.vout - b.vout)
outputs.sort((a, b) => a.value - b.value || compare(a.script, b.script, MIN(a.script.length, b.script.length)))
``` 

Where `compare` is just a standard lexicographical comparison function, of which there are numerous examples you could give beyond a full python implementation as listed above.

* `std::lexicographical_compare` in [C++]* (http://www.cplusplus.com/reference/algorithm/lexicographical_compare/)
* `cmp` in Python 2.7
* `memcmp` in C

Very easy to find,  very easy to [research](https://en.wikipedia.org/wiki/Lexicographical_order).
Almost always pointless to re-implement,  or even give a pseudo-algorithm for.

I'd really like to see this BIP get through, but as an open source implementer, it took me more time to decipher the paragraphs behind this BIP than it did to write the [~25 line `npm` module](https://github.com/bitcoinjs/bip69/blob/master/index.js).

**Comment by kristovatlas** [comment]
Okey doke. I don't think this matters at all and is a waste of my time. If you want to fork and make a pull request to my repo, I will promise to review it promptly, however.

**Comment by gmaxwell** [comment]
@dcousens  Any thoughts on if you will try a revision. What you wrote there looks like a reasonable simplification which could be merged with Kristov's work.

**Comment by dcousens** [comment]
@gmaxwell waiting on https://github.com/kristovatlas/bips/pull/1

**Comment by kristovatlas** [comment]
@gmaxwell merged @dcousens's edits. I think this PR is ready to be merged.

**Comment by dcousens** [comment]
ACK

**Comment by kristovatlas** [comment]
Thanks for the editing help, @dcousens

**Comment by kristovatlas** [comment]
@laanwj ping

**Comment by kristovatlas** [comment]
@laanwj ping

**Comment by luke-jr** [comment]
README.mediawiki and bip-0069/bip-0069_examples.py are missing EOF newlines. Anything else needed before merging this draft?

**Comment by kristovatlas** [comment]
done

**Comment by MarcoFalke** [comment]
Created two trivial follow up cleanups:  #203 and  #204. Feedback is welcome.

**Pull request merged**